lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+            return self._private[dir_prop]
+        end
+    end
+
+    -- Non-directional options
+    grid["set_"..prop] = function(self, value)
+        if self._private["horizontal_"..prop] ~= value or self._private["vertical_"..prop] ~= value then
+            self._private["horizontal_"..prop] = value
+            self._private["vertical_"..prop] = value
+            self:emit_signal("widget::layout_changed")
+        end
+    end
+    grid["get_"..prop] = function(self)
+        if self._private.orientation == "x" then
+            return self._private["horizontal_"..prop]
+        elseif self._private.orientation == "y" then
+            return self._private["vertical_"..prop]
+        end
+    end
+end
+
+
+-- Return two tables of the fitted sizes of the rows and columns
+-- @return table,table Tables of row heights and column widths
+local function get_matrix_sizes(self, context, orig_width, orig_height)
+    local matrix = self._private.matrix
+    local rows_size = {}
+    local cols_size = {}
+
+    -- Set the row and column sizes to the minimum value
+    for i = 1,self._private.num_rows do
+        rows_size[i] = self._private.min_rows_size
+    end
+    for j = 1,self._private.num_cols do
+        cols_size[j] = self._private.min_cols_size
+    end
+
+    -- Calculate cell sizes
+    local line
+    for row = 1, self._private.num_rows do
+        line = matrix[row]
+        if type(line) == "table" then -- check that the row exists
+            for col = 1, self._private.num_cols do
+                local data = matrix[row][col]
+                if data then
+                    -- fit the widget to calculate the cell size
+                    local w, h = base.fit_widget(self, context, data.widget, orig_width, orig_height)
+                    h = math.max( self._private.min_rows_size, h / data.row_span )
+                    w = math.max( self._private.min_cols_size, w / data.col_span )
+                    -- update the current row and column maximum size
+                    if h > rows_size[row] then rows_size[row] = h end
+                    if w > cols_size[col] then cols_size[col] = w end
+                end
+            end
+        end
+    end
+    return rows_size, cols_size
+end
+
+--- Fit the grid layout into the given space
+-- @param context The context in which we are fit.
+-- @param orig_width The available width.
+-- @param orig_height The available height.
+function grid:fit(context, orig_width, orig_height)
+    local width, height = orig_width, orig_height
+    local used_width_max = 0
+    local used_height_max = 0
+
+    -- fit matrix cells
+    local rows_size, cols_size = get_matrix_sizes(self, context, width, height)
+
+    -- compute the width
+    if self._private.horizontal_fixed_size then
+        -- all the columns have the same size
+        used_width_max = #cols_size * max_value(cols_size) + (#cols_size - 1) * self._private.horizontal_spacing + 2 * self._private.horizontal_padding
```

Comment:
Why does this widget implement its own padding, by the way? A `margin` widget around the grid could just do so as well.